In [0]:
import tensorflow as tf
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.manifold import TSNE
import scipy.spatial.distance
from scipy import stats

plt.set_cmap('gray')

np.random.seed(1337)

<Figure size 432x288 with 0 Axes>

In [0]:
def data_preparation():
  #Trainings- und Testdaten
  global digits_train, labels_train
  mnist_train = tf.keras.datasets.mnist.load_data()[0]
  digits_train, labels_train = mnist_train[0], mnist_train[1]

  #Anzahl der Samples in train und test
  global n_train
  n_train = digits_train.shape[0]

  #Anzahl der Pixel
  m = digits_train.shape[1]**2

  #Vektor aus Matrizen erstellen -> digits flat
  digits_train = digits_train.reshape([n_train, m])

In [0]:
def kmeans(n_iterations):
  print("KMEANS für", n_iterations, "Iterationen")
  #1000 Sampledaten aus den 60000 Trainingsdaten
  n_sample = 60000
  idx = np.random.randint(n_train, size=n_sample)
  global x_sample, y_sample
  x_sample = digits_train[idx]
  y_sample = labels_train[idx]
  #Anzahl Cluster
  global k
  k = 10

  # init centroids
  # 10 random Centroids finden
  idx_centroid = np.random.randint(n_sample, size=k)
  #Centroids in eigene Variable speichern
  centroids = x_sample[idx_centroid]

  idx_centroid

  # init centroids
  #range für iterations anpassen
  global labels
  for j in range(n_iterations):
      
      # calc distances
      dists = scipy.spatial.distance.cdist(x_sample, centroids, metric='euclid')
      
      # label according to nearest centroids
      
      labels = np.argmin(dists, axis=1)
      
      # get new centroids
      centroid_list = [x_sample[labels == i].mean(0) for i in range(k)]
      centroids = np.stack(centroid_list)

In [0]:
def cluster_output():
  for i in range(k):
    size = y_sample[labels == i].shape[0]
    count_numbers = np.bincount(y_sample[labels == i])
    number_idx = count_numbers.argmax()
    correct = count_numbers[number_idx]
    percentage = (correct / size) * 100

    print("Clustergröße: ", size)
    print("Verteilung der Zahlen", count_numbers)
    print("Index / Zahl der meisten Zahlen", number_idx)
    print("Richtig zugeordnet", correct)
    print("Cluster " + str(i) + " besteht zu " + str(round(percentage, 2)) + " % aus der Zahl " + str(number_idx))
    print("___________________________________________")

In [0]:
#Methoden werden hier aufgerufen

In [0]:
data_preparation()
kmeans(10)
cluster_output()
print("====================")

kmeans(100)
cluster_output()
print("====================")

kmeans(1000)
cluster_output()

KMEANS für 10 Iterationen
Clustergröße:  4491
Verteilung der Zahlen [4266    0   41   20    1   48   57   10   19   29]
Index / Zahl der meisten Zahlen 0
Richtig zugeordnet 4266
Cluster 0 besteht zu 94.99 % aus der Zahl 0
___________________________________________
Clustergröße:  6921
Verteilung der Zahlen [ 206   15  305 1878    3 1307   28    3 3136   40]
Index / Zahl der meisten Zahlen 8
Richtig zugeordnet 3136
Cluster 1 besteht zu 45.31 % aus der Zahl 8
___________________________________________
Clustergröße:  6489
Verteilung der Zahlen [1036    2  655  307  320 1575 2068   16  466   44]
Index / Zahl der meisten Zahlen 6
Richtig zugeordnet 2068
Cluster 2 besteht zu 31.87 % aus der Zahl 6
___________________________________________
Clustergröße:  7001
Verteilung der Zahlen [   7   12   54   29 1444  131    1 3131  261 1931]
Index / Zahl der meisten Zahlen 7
Richtig zugeordnet 3131
Cluster 3 besteht zu 44.72 % aus der Zahl 7
___________________________________________
Clustergröße: 

# **Antwort**
Basierend auf 60000 Samples.

**What is the majority class of each cluster and what is the percentage of the majority class in each cluster??**

**Does each number have a cluster? If not, which hasn’t?**

*KMEANS: 10 Iterationen*

Cluster 0 besteht zu 94.99 % aus der Zahl 0

Cluster 1 besteht zu 45.31 % aus der Zahl 8

Cluster 2 besteht zu 31.87 % aus der Zahl 6

Cluster 3 besteht zu 44.72 % aus der Zahl 7

Cluster 4 besteht zu 56.96 % aus der Zahl 1

Cluster 5 besteht zu 44.15 % aus der Zahl 4

Cluster 6 besteht zu 70.41 % aus der Zahl 1

Cluster 7 besteht zu 49.31 % aus der Zahl 3

Cluster 8 besteht zu 48.63 % aus der Zahl 6

Cluster 9 besteht zu 32.26 % aus der Zahl 9

Hier haben die Zahlen 2 und 5 kein Cluster. Die Zahlen 1 und 6 haben dafür zwei Cluster bekommen. Wie man am prozentualen Anteil erkennen kann sind die Mehrheit im Cluster nicht immer eindeutig. Nur für die Zahl 0 haben wir eine Cluster Mehrheit von 94.99% erreicht. 

-----

*KMEANS: 100 Iterationen*

Cluster 0 besteht zu 40.28 % aus der Zahl 7

Cluster 1 besteht zu 59.37 % aus der Zahl 1

Cluster 2 besteht zu 49.87 % aus der Zahl 8

Cluster 3 besteht zu 84.86 % aus der Zahl 6

Cluster 4 besteht zu 50.67 % aus der Zahl 3

Cluster 5 besteht zu 35.29 % aus der Zahl 9

Cluster 6 besteht zu 87.01 % aus der Zahl 2

Cluster 7 besteht zu 40.59 % aus der Zahl 4

Cluster 8 besteht zu 65.60 % aus der Zahl 1

Cluster 9 besteht zu 94.63 % aus der Zahl 0

Hier hat die Zahl 5 erneut kein Cluster. Dies war schon bei 10 Iterationen der Fall. Ansonsten haben alle Zahlen ein Cluster. Die Zahl 0 hat erneut eine deutliche Cluster Mehrheit. Die Zahlen 2 und 6 könnten ebenfalls gute Werte erreichen.

----

*KMEANS: 1000 Iterationen*

Cluster 0 besteht zu 91.80 % aus der Zahl 2

Cluster 1 besteht zu 94.41 % aus der Zahl 0

Cluster 2 besteht zu 30.44 % aus der Zahl 6

Cluster 3 besteht zu 40.90 % aus der Zahl 4

Cluster 4 besteht zu 53.65 % aus der Zahl 3

Cluster 5 besteht zu 44.68 % aus der Zahl 7

Cluster 6 besteht zu 67.74 % aus der Zahl 1

Cluster 7 besteht zu 88.38 % aus der Zahl 6

Cluster 8 besteht zu 34.78 % aus der Zahl 7

Cluster 9 besteht zu 59.98 % aus der Zahl 8

Und auch bei 1000 Iterationen konnte für die 5 kein Mehrheitscluster gefunden werden. Das bedeutet in allen drei Fällen hat die fünf kein Mehrheitscluster bekommen. Dazu wurde hier für die 9 keine Mehrheit in einem Cluster gefunden. Die Zahlen 6 und 7 haben dafür zwei Cluster.

Die genaue Verteilung und Clustergröße ist oben strukturiert zu finden. Das Array Verteilung der Zahlen (im Code count_numbers) ist von 0-9 struktutiert. [0] ist also die Zahl 0.